### ISSUE TITLE
Failed to execute ban

### ISSUE THREAD

**Issue by Git-Forked** [fail2ban/fail2ban#3760]
### fail2ban.log
```
2024-06-06 08:19:38,862 fail2ban.actions        [761]: NOTICE  [manual] Ban 103.131.71.0/24
2024-06-06 08:19:38,872 fail2ban.utils          [761]: ERROR   7f784580e8e0 -- exec: ipset -exist add f2b-manual 103.131.71.0/24 timeout 0
2024-06-06 08:19:38,872 fail2ban.utils          [761]: ERROR   7f784580e8e0 -- stderr: 'ipset v7.17: Hash is full, cannot add more elements'
2024-06-06 08:19:38,872 fail2ban.utils          [761]: ERROR   7f784580e8e0 -- returned 1
2024-06-06 08:19:38,880 fail2ban.utils          [761]: ERROR   7f784580e8e0 -- exec: ipset -exist add f2b-manual 103.131.71.0/24 timeout 0
2024-06-06 08:19:38,881 fail2ban.utils          [761]: ERROR   7f784580e8e0 -- stderr: 'ipset v7.17: Hash is full, cannot add more elements'
2024-06-06 08:19:38,881 fail2ban.utils          [761]: ERROR   7f784580e8e0 -- returned 1
2024-06-06 08:19:38,881 fail2ban.actions        [761]: ERROR   Failed to execute ban jail 'manual' action 'firewallcmd-ipset' info 'ActionInfo({'ip': '103.131.71.0/24', 'family': 'inet4', 'fid': <function Actions.ActionInfo.<lambda> at 0x7f7845749300>, 'raw-ticket': <function Actions.ActionInfo.<lambda> at 0x7f7845749a80>})': Error banning 103.131.71.0/24
```

### /etc/fail2ban/jail.d/manual.local
```
[manual]
enabled = true
port = http,https
bantime = -1
```

### /etc/fail2ban/action.d/manual.local
```
[Definition]
actionstart = ipset -exist create <ipmset> hash:net timeout <default-ipsettime> maxelem 1000000 <familyopt>
```
**(note:   hash:net seems to be ignored and the jail gets hash:ip instead) [would prefer this one jail to use hash:net for banning subnets in CIDR format]**

### List of subnets that should be banned in this jail
```
103.131.71.0/24
104.131.0.0/16
104.152.52.0/24
106.75.0.0/16
107.170.0.0/16
110.154.0.0/16
110.155.0.0/16
111.7.0.0/16
117.151.0.0/16
117.153.135.0/24
120.227.0.0/16
120.69.0.0/16
123.0.0.0/8
124.119.0.0/16
128.199.0.0/16
134.122.0.0/16
134.209.0.0/16
137.184.0.0/16
138.197.0.0/16
138.68.0.0/16
13.91.0.0/16
139.211.0.0/16
139.213.0.0/16
139.59.0.0/16
141.98.0.0/16
142.93.0.0/16
143.110.0.0/16
146.190.0.0/16
146.19.0.0/16
147.78.103.0/24
148.153.0.0/16
149.102.242.0/24
149.50.0.0/16
151.237.177.0/24
152.32.0.0/16
152.42.0.0/16
156.239.0.0/16
157.230.0.0/16
157.245.0.0/16
159.203.0.0/16
159.65.0.0/16
161.35.0.0/16
161.97.0.0/16
162.142.125.0/24
162.243.0.0/16
164.92.0.0/16
165.154.0.0/16
165.22.0.0/16
165.227.0.0/16
167.248.133.0/24
167.71.0.0/16
167.94.0.0/16
170.64.0.0/16
172.0.0.0/8
175.22.0.0/16
176.97.113.0/24
178.128.0.0/16
178.62.0.0/16
179.43.0.0/16
182.201.0.0/16
182.204.0.0/16
184.105.0.0/16
185.0.0.0/8
188.166.0.0/16
192.241.0.0/16
192.42.0.0/16
193.177.182.0/24
194.48.251.0/24
194.59.0.0/16
195.1.144.0/24
198.199.0.0/16
198.235.0.0/16
199.45.0.0/16
20.118.0.0/16
206.168.34.0/24
206.189.0.0/16
209.141.0.0/16
209.38.0.0/16
216.218.206.0/24
216.244.66.0/24
218.31.0.0/16
222.81.0.0/16
223.113.128.0/24
223.84.0.0/16
223.88.0.0/16
223.89.0.0/16
23.137.104.0/24
23.80.153.0/24
23.81.0.0/16
27.20.0.0/16
3.0.0.0/8
31.220.0.0/16
35.203.0.0/16
35.216.0.0/16
36.0.0.0/8
39.0.0.0/8
41.0.0.0/8
4.151.218.0/24
42.202.0.0/16
4.246.246.0/24
42.83.147.0/24
43.0.0.0/8
45.0.0.0/8
46.101.0.0/16
47.0.0.0/8
49.114.0.0/16
51.0.0.0/8
5.0.0.0/8
52.0.0.0/8
64.0.0.0/8
65.49.0.0/16
78.153.0.0/16
79.110.0.0/16
80.0.0.0/8
8.0.0.0/8
83.147.0.0/16
84.247.0.0/16
84.54.0.0/16
87.121.0.0/16
91.223.0.0/16
91.92.0.0/16
92.118.0.0/16
93.115.0.0/16
94.156.0.0/16
```

**Comment by Git-Forked** [comment]
Possibly related bug of firewalld:  

```
What appears to happen when you try to put in something like 1.2.3.4/8 
is that the entire address space attempts to be expanded. 

# firewall-cmd --permanent --new-ipset=test --type=hash:ip
success

# firewall-cmd --permanent --ipset=test --add-entry=1.2.3.4/8
success

# firewall-cmd --reload
Error: COMMAND_FAILED: '/usr/sbin/ipset restore' failed: ipset v7.1: Error in line 3: Hash is full, cannot add more elements
```
source: https://bugzilla.redhat.com/show_bug.cgi?id=1777909

**Comment by sebres** [comment]
> `ipset v7.17: Hash is full, cannot add more elements`

It is clearly a "bug" of ipset (or your action possibly), so my question is - why you're trying to post it here?

Anyway, it is clearly the restriction of itpset - the error simply means that the ipset elements limit is exceeded.
We have nothing to do with that.
However related to [this article](https://cloudlinux.zendesk.com/hc/en-us/articles/360017209480-Ipset-max-elements-error-Hash-is-full-cannot-add-more-elements) (did not verified that):
*The ipset size is hardcoded in the Imunify360 source code and currently, it is equal to a 100K IPs limit.*

So your attempt to increase it with `maxelem 1000000` may fail.
However, I guess it may depend on kernel version or some kernel module...

By the way, I don't think you'll really ban 1M subnets... It is too heavy and looks superfluous for `hash:net`.

> hash:net seems to be ignored and the jail gets hash:ip instead) [would prefer this one jail to use hash:net for banning subnets in CIDR format

I don't think so. What do you see with `ipset list f2b-manual`?

If I try it with your list, I see this after all:
```bash
# while read ip; do; echo "$ip"; fail2ban-client set manual banip "$ip"; done </tmp/net-list.txt
# ipset list f2b-manual
Name: f2b-manual
Type: hash:net
Revision: 7
Header: family inet hashsize 1024 maxelem 1000000 timeout 0 bucketsize 12 initval 0xa9941845
Size in memory: 7944
References: 0
Number of entries: 124
Members:
... a lot of subnets ...
```

If you mean fail2ban makes something wrong (e. g. with actionstart) please provide an excerpt for:
```
fail2ban-client -d | grep manual
```

**Comment by sebres** [comment]
If you previously changed
```diff
- actionstart = ipset -exist create <ipmset> hash:ip ...
+ actionstart = ipset -exist create <ipmset> hash:net ...
```
in the config, please ensure you **restart** (not just reload) fail2ban or the jail, in order to recreate the action and reban everything.
Without restart, fail2ban would hold previously created action (and therefore doesn't invoke actionstart, doesn't reban), so ipset may remain `hash:ip`. And then indeed ipset would expand subnet to a lot of IPs.

**Comment by Git-Forked** [comment]
### ipset list f2b-manual --terse
```
Name: f2b-manual
Type: hash:ip
Revision: 5
Header: family inet hashsize 1048576 maxelem 1000000 timeout 0 bucketsize 12 initval 0x142aae92
Size in memory: 46316208
References: 1
Number of entries: 1000000
```

**Comment by sebres** [comment]
> `'actionstart', 'ipset -exist create <ipmset> hash:ip ...`

As you can see, the action config has still original type of set.
Unfortunately, we haven't currently a parameter for ipset-based actions to set the type of ipset only... 

And setting of `actionstart` in jail (as direct parameter of jail) would be not correct...
I don't know where you had set it, but there are basically only 3 possibilities at the moment:
- either to create a derivative action, including the action you want to use, and there overwrite `actionstart` or a part responsible for the ipset type inside corresponding section. And then set name of this action as new banaction for jail.
- almost the same as above but make a duplicate of action, where you'd overwrite the needed stuff.
- or supply `actionstart` as parameter **for `action`** in the jail.

Written from mobile phone, so too unhandy to provide an example... Please ping me if you'd need one, I'll provide it later.

**Comment by Git-Forked** [comment]
I set the directive
```
[Definition]
actionstart = ipset -exist create <ipmset> hash:net timeout <default-ipsettime> maxelem 1000000 <familyopt>
```
 in the /etc/fail2ban/action.d/manual.local

**Comment by sebres** [comment]
> I set the directive ... in the /etc/fail2ban/action.d/manual.local

Well, this looks like 2 errors to me (if you did only that):
- the action missing everything else, because neither have an include to original action, nor other directives expected for work
- probably you did not set it as `banaction` in jail.

Ok, below is an example for stock `iptables-ipset`, to rewrite this:
https://github.com/fail2ban/fail2ban/blob/304c3cd56693c9ba90ace8b11c656ac2912648ca/config/action.d/iptables-ipset.conf#L27-L28

#### in the action.d/manual.local
```ini
[INCLUDES]
before = iptables-ipset.conf

[Definition]
actionstart = ipset -exist create <ipmset> hash:net timeout <default-ipsettime> maxelem <maxelem> <familyopt>
              <_ipt_add_rules>
```
(if your original action doesn't have `<maxelem>` parameter, simply use 1000000 like you did, although I doubt you'd ever ban 1M subnets).

#### in the jail:
```ini
[manual]
enabled = true
banaction = manual
...
```
(for `firewallcmd-ipset` action it would be similar: change the include and adjust ipset creation in related sections).

However a bit ugly is if tomorrow the original action would change in an incompatible way (e. g. `_ipt_add_rules` doesn't exist anymore or `actionstart` becomes more commands), this may fail after upgrade to newer version.

Alternative would be something like that:
#### in the action.d/manual.local
```ini
[INCLUDES]
before = iptables-ipset.conf

[Definition]
actionstart = ipset -exist create <ipmset> hash:net timeout <default-ipsettime> maxelem <maxelem> <familyopt>
              %(known/actionstart)s
```
where it'd firstly create the set with your parameters and then evaluate original `actionstart`, which also trying to create the set firstly (but would do nothing because of `-exist`) but hereafter would evaluate further commands... and since `actionstart` executes only once per ip-family, it looks like good workaround to me.

**Important:** don't forget to restart (not reload) the jail after all.

**Comment by sebres** [comment]
25335268276184b5339bb066f19e178da07e3366 implements now new parameter `ipsettype` to support this...
One can simply use this action instead (if the fail2ban version is at least v.1.0).

**Important:** for `action.d/firewallcmd-ipset.conf` old parameter `ipsettype` initially added in #2620 for an ipset-backend renamed now to `ipsetbackend`, because new parameter `ipsettype` will be used for the real set type

**Comment by melroy89** [comment]
> [2533526](https://github.com/fail2ban/fail2ban/commit/25335268276184b5339bb066f19e178da07e3366) implements now new parameter `ipsettype` to support setting of set type... One can simply use this action instead (if the fail2ban version is at least v.1.0).
> 
> For instance:
> 
> ```ini
> # for multiport:
> banaction = iptables-ipset[type=multiport, ipsettype=hash:net]
> # for allports:
> banaction = iptables-ipset[type=allports, ipsettype=hash:net]
> ```
> 
> **Important:** for `action.d/firewallcmd-ipset.conf` old parameter `ipsettype` initially added in #2620 for an ipset-backend renamed now to `ipsetbackend`, because new parameter `ipsettype` will be used for the real set type

What is the new max limit of the above banactions using `ipsetype=hash:net`? Or should I use `firewallcmd-ipset[ipsetbackend=firewalld]`?

**Comment by sebres** [comment]
What is the new max limit of the above banactions using `ipsetype=hash:net`?

The type `ipsetype=hash:net` is not about the limit, it is about how ipset stores net-range (e. g. test-net `192.0.2.0/24`), in case of default `ipsetype=hash:ip` ipset (if supported) would store **256 entries** `192.0.2.0 .. 192.0.2.255`, in case of `ipsetype=hash:net` it stores only one (particularly the subnet `192.0.2.0/24` itself as a **single entry**).
 The limit remains the same, just it occupies fewer entries (in case of subnet). For banning of normal IP-addresses it is useless. However setting of wrong `ipsetype` (e. g. `hash:ip` by banning of subnets) can cause exploding of the entries number and therefore faster reach the limit, but how fast would depend on the sizes of subnets you'd add (ban).

Regarding the limit: there was different PR (#3564) and parameter `maxelem` (by default 65536), and affects things like hashsize or bucketsize.
But again, the settings `ipsetype` and `maxelem` are orthogonal to each other... but in case of `hash:ip` it means `maxelem` particular IP-addresses, in case of `hash:net` it means `maxelem` subnets.

This example as diff illustrates what I mean (maxelem is intentionally so low, to show that `192.0.2.0/24` would even not fit into `hash:ip` with 256 addresses by "exploding"):
```diff
- # ipset create f2b-test hash:ip maxelem 20
+ # ipset create f2b-test hash:net maxelem 20

  # ipset add f2b-test 192.0.2.0/24 && echo ok
- ipset v7.17: Hash is full, cannot add more elements
+ ok

  # ipset list f2b-test
  Name: f2b-test
- Type: hash:ip
+ Type: hash:net
  ...
- Number of entries: 20
+ Number of entries: 1
  Members:
- 192.0.2.2
- 192.0.2.1
- 192.0.2.7
- 192.0.2.0
- 192.0.2.9
- 192.0.2.4
- 192.0.2.3
- 192.0.2.6
- 192.0.2.5
- 192.0.2.8
- ...
+ 192.0.2.0/24

  # ipset destroy f2b-test
```
Also important: the type `hash:net` is basically more correct if you really ban subnets, because by `hash:ip` two crossing but different subnets (one gets banned another unbanned) would produce a certain "conflict" (because the IPs are exploded and by unban the crossing part would be removed from set hash), whereas by `hash:net` they remains separate entries.
For the example above, unban of `192.0.2.8/29` by `hash:ip` will remove from set the range `192.0.2.8 .. 192.0.2.15`.

**Comment by melroy89** [comment]
Ah, this helps me a lot! Thank you so much for clarifying. This is exactly what I needed, since I was indeed banning large amount of subnet.

Just what happens if I also ban 3 or 4 single IPs with `hash:net`?  That still works fine right? 99% are subnets but maybe a couple aren't. Or should I then better create two jails..

**Comment by sebres** [comment]
You are welcome...

> Just what happens if I also ban 3 or 4 single IPs with `hash:net`? That still works fine right?

Sure. Since an IP is also a subnet (e. g. `192.0.2.1` is the same than `192.0.2.1/32` or `2001:db8::1` is equal `2001:db8::1/128`), just they contain only one address.

**Comment by melroy89** [comment]
True!

**Comment by melroy89** [comment]
One small follow-up question I have now.. I did change the config in fail2ban by setting:

```diff
[manual]
enabled = true
port    = http,https
-banaction = iptables-ipset-proto6-allports
+banaction = iptables-ipset[type=allports, ipsettype=hash:net]
```

However: `ipset list f2b-manual | less` still shows me `Type: hash:ip`.. 

I tried to restart fail2ban. But I can't get fail2ban to change the ipset type I think. It is advised to just manually remove `f2b-manual` list from ipset? Is that the way?

**Comment by melroy89** [comment]
I'm despite now. So I also installed the latest deb: https://github.com/fail2ban/fail2ban/releases/download/1.1.0/fail2ban_1.1.0-1.upstream1_all.deb. So now I'm running Fail2Ban v1.1.0.1.

**Comment by sebres** [comment]
Either it doesn't upgraded the action for some reason... What do you see in your /etc/fail2ban/action.d/iptables-ipset.conf? It must be:
https://github.com/fail2ban/fail2ban/blob/89970d2e3e298d18abd3698317483828550cbe43/config/action.d/iptables-ipset.conf#L27

Or the ipset get preserved (can't be deleted by restart, e.g. due to some dependency) so it'd be not recreated. Just check it after fail2ban stop (and remove the set with `ipset` command manually).

**Comment by melroy89** [comment]
> Either it doesn't upgraded the action for some reason... What do you see in your /etc/fail2ban/action.d/iptables-ipset.conf? It must be:
> 
> https://github.com/fail2ban/fail2ban/blob/89970d2e3e298d18abd3698317483828550cbe43/config/action.d/iptables-ipset.conf#L27
> 
> Or the ipset get preserved (can't be deleted by restart, e.g. due to some dependency) so it'd be not recreated. Just check it after fail2ban stop (and remove the set with `ipset` command manually).

Interesting.. As I said, I did install the latest deb manually (`apt install ./fail2ban_1.1.0-1.upstream1_all.deb`), which also updated this  `iptables-ipset.conf` file actually.

However, as you corrctly pointed out the file `/etc/fail2ban/action.d/iptables-ipset.conf` seems not reflect the actual changes in your repository conf file:

```
# Option:  actionstart
# Notes.:  command executed on demand at the first ban (or at the start of Fail2Ban if actionstart_on_demand is set to false).
# Values:  CMD
#
actionstart = ipset -exist create <ipmset> hash:ip timeout <default-ipsettime> maxelem <maxelem> <familyopt>
              <_ipt_add_rules>
```

Your debian installation is wrongly packaged. Look at the deb file content:

![image](https://github.com/user-attachments/assets/c93f767e-fa3a-4a1c-b821-1830a318cfee)

The `hash:ip` is still hardcoded in this conf file and I change manually this file now to reflect the config you mentioned.

You might want to consider repacking the deb file?
